docs/source/llms/custom-pyfunc-for-llms/notebooks/custom-pyfunc-advanced-llm.ipynb

```diff
@@ -0,0 +1,461 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Serving LLMs with MLflow: Leveraging Custom PyFunc\n",
+    "\n",
+    "\n",
+    "#### Introduction\n",
+    "\n",
+    "In this tutorial, we'll explore how to save custom cutting-edge Large Language Models (LLMs) using MLflow. Specifically, we'll delve into the intricacies of a situation where the default MLflow 'transformers' flavor does not provide direct support for our model type and its dependencies. This necessitates the creation of a custom `pyfunc` to ensure seamless model deployment. \n",
+    "\n",
+    "Through this tutorial, we aim to provide you with:\n",
+    "\n",
+    "- An understanding of why certain models might need custom `pyfunc` definitions.\n",
+    "\n",
+    "- A walk-through of creating a custom `pyfunc` to handle model dependencies and interface data.\n",
+    "\n",
+    "- Insight into how a custom `pyfunc` can offer a simplified interface to end-users in a deployed environment.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### The Challenge with Default Implementations\n",
+    "\n",
+    "MLflow's `transformers` flavor provides a standardized way to handle models from the HuggingFace Transformers library. However, not all models or configurations might fit neatly into this standardized format. \n",
+    "\n",
+    "In our scenario, the model cannot use the default `pipeline` type due to certain incompatibilities. This poses a challenge: how do we ensure that our model can be saved, loaded, and served using MLflow, given these constraints?\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "#### The Power of Custom PyFunc\n",
+    "\n",
+    "The solution lies in MLflow's ability to define custom `pyfunc`. By creating a custom `pyfunc`, we can:\n",
+    "\n",
+    "- Define how the model loads its dependencies.\n",
+    "\n",
+    "- Customize the inference process.\n",
+    "\n",
+    "- Manipulate interface data to create specific inputs for the model.\n",
+    "\n",
+    "Let's dive into the code to see this in action.\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Important Considerations Before Proceeding\n",
+    "\n",
+    "#### Hardware Recommendations\n",
+    "This guide demonstrates the usage of a particularly large and intricate Large Language Model (LLM). Given its complexity:\n",
+    "\n",
+    "- **GPU Requirement**: It's **strongly advised** to run this example on a system with a CUDA-capable GPU that possesses at least 64GB of VRAM.\n",
```

Comment:
Do we really need `64GB of VRAM` to demonstrate how we can use custom pyfunc with LLM?